Fix FILE resource leak

nmcclatchey opened the pull request:
Fixes resource leaks in `npy_load()` and `npz_load()` that would occur if an exception were to be thrown during load. Because NPY files can be constructed deliberately to trigger such exceptions, addressing this leak is vital for cases where programs may load user-supplied NPY files and attempt recovery on error.

nmcclatchey committed:
Close FILE pointer on exception in npz_load

Fixes issue #53, which notes that an exception thrown when loading an npz file would cause an open FILE pointer to be leaked. This bug could be triggered intentionally (via specially-crafted NPZ files), so fixing it is a priority.

nmcclatchey committed:
Close FILE pointer on exception in npy_load

Fixes issue #52, which notes that an exception thrown when loading an npy file would cause an open FILE pointer to be leaked. This bug could be triggered intentionally (via specially-crafted NPY files), so fixing it is a priority.

nmcclatchey commented:
Note: Resolves #52 and #53 .

s-trinh commented:
Thanks for the code. I have used it with the following changes:

```cpp
  FILE *fp;
  AutoCloser() : fp(nullptr) { }
  ~AutoCloser()
  {
    if (fp != nullptr) {
      fclose(fp);
    }
  }
};
```
I have added the constructor, otherwise my compiler complains about `fp` not being initialized.
The `fp != nullptr` is probably useless but it looks like sending `NULL` (so also `nullptr`?) to `fclose` is undefined behavior.

nmcclatchey committed:
Apply feedback

Addresses a note about potential undefined behavior on `nullptr` (and a false-positive compiler complaint about uninitialized values).

nmcclatchey commented:
Hi @s-trinh. I like your feedback, and have applied it (less one redundant write) to the PR.
